src/test/app/Delegate_test.cpp

```diff
@@ -0,0 +1,1308 @@
+            env.close();
+
+            // now bob create another Delegate, he does not have
+            // enough reserve
+            env(delegate_set::delegateSet(bob, carol, {"Payment"}),
+                ter(tecINSUFFICIENT_RESERVE));
+        }
+
+        // test reserve when sending transaction on behalf of other account
+        {
+            Env env(*this, features);
+            Account alice{"alice"};
+            Account bob{"bob"};
+
+            env.fund(drops(env.current()->fees().accountReserve(1)), alice);
+            env.fund(drops(env.current()->fees().accountReserve(2)), bob);
+            env.close();
+
+            // alice gives bob permission
+            env(delegate_set::delegateSet(alice, bob, {"DIDSet", "DIDDelete"}));
+            env.close();
+
+            // bob set DID on behalf of alice, but alice does not have enough
+            // reserve
+            env(did::set(alice),
+                did::uri("uri"),
+                delegate(bob),
+                ter(tecINSUFFICIENT_RESERVE));
+
+            // bob can set DID for himself because he has enough reserve
+            env(did::set(bob), did::uri("uri"));
+            env.close();
+        }
+    }
+
+    void
+    testFee(FeatureBitset features)
+    {
+        testcase("test fee");
+        using namespace jtx;
+
+        Env env(*this, features);
+        Account alice{"alice"};
+        Account bob{"bob"};
+        Account carol{"carol"};
+        env.fund(XRP(10000), alice, carol);
+        env.fund(XRP(1000), bob);
+        env.close();
+
+        {
+            // Fee should be checked before permission check,
+            // otherwise tecNO_PERMISSION returned when permission check fails
+            // could cause context reset to pay fee because it is tec error
+            auto aliceBalance = env.balance(alice);
+            auto bobBalance = env.balance(bob);
+            auto carolBalance = env.balance(carol);
+
+            env(pay(alice, carol, XRP(100)),
+                fee(XRP(2000)),
+                delegate(bob),
+                ter(terINSUF_FEE_B));
+            env.close();
+            BEAST_EXPECT(env.balance(alice) == aliceBalance);
+            BEAST_EXPECT(env.balance(bob) == bobBalance);
+            BEAST_EXPECT(env.balance(carol) == carolBalance);
+        }
+
+        env(delegate_set::delegateSet(alice, bob, {"Payment"}));
+        env.close();
+
+        {
+            // Delegate pays the fee
+            auto aliceBalance = env.balance(alice);
+            auto bobBalance = env.balance(bob);
+            auto carolBalance = env.balance(carol);
+
+            env(pay(alice, carol, XRP(100)), fee(XRP(10)), delegate(bob));
+            env.close();
+            BEAST_EXPECT(env.balance(alice) == aliceBalance - XRP(100));
+            BEAST_EXPECT(env.balance(bob) == bobBalance - XRP(10));
+            BEAST_EXPECT(env.balance(carol) == carolBalance + XRP(100));
```

Comment:
```suggestion
            auto const sendAmt = XRP(100);
            auto const feeAmt = XRP(10);
            env(pay(alice, carol, sendAmt), fee(feeAmt), delegate(bob));
            env.close();
            BEAST_EXPECT(env.balance(alice) == aliceBalance - sendAmt);
            BEAST_EXPECT(env.balance(bob) == bobBalance - feeAmt);
            BEAST_EXPECT(env.balance(carol) == carolBalance + sendAmt);
```

should define variables